Improving neurodamus build w support for PythonND

ferdonline opened the pull request:
This change implements:

- Improved build, avoiding copying files by completely overriding the `do_stage` step. We split the build into "build" and "install", where build only builds the special, and install puts everything in a nice place without duplicating stuff. The new install structure is:
  - `$prefix/bin/special`
  - `$prefix/lib/modc` (mods c source)
  - `$prefix/lib/hoclib` (symlink to hoclib)
  - `$prefix/lib/modlib` (symlink to modlib)
  - `$prefix/lib/libnrnmech.so` (moved from ARCH if exists - compiled in shared lib mode)
  - `$prefix/python` (symlink to python srcs, if exists)
 - Support for a python variant, declaring dependencies
 - Rename of variant syn2 to syntool

ferdonline committed:
Refactoring neurodamus building for python support

pramodk commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -34,9 +35,10 @@ class Neurodamus(NeurodamusBase):
     variant('coreneuron', default=True, description="Enable CoreNEURON Support")
     variant('profile', default=False, description="Enable profiling using Tau")
-    variant('syn2', default=False, description="Enable Synapsetool reader")
+    variant('syntool', default=False, description="Enable Synapsetool reader")
```
dont you want to make syntool as default?

pramodk commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -53,28 +55,37 @@ class Neurodamus(NeurodamusBase):
+    depends_on("py-setuptools", when='+python')
+    depends_on('py-enum34', when='^python@2.4:2.7.999,3.1:3.3.999')
+    depends_on("py-lazy-property", when='+python')
+
```
If I am not mistaken python extension packages needs to be `type=('build', 'run'))` in dependencies. (they are not needed as link dependency)

pramodk commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -83,41 +94,78 @@ def syn2_dep_libs(self):
+        if os.path.exists(arch + "/.libs/libnrnmech.so"):
+            shutil.move(arch + "/.libs/libnrnmech.so", prefix.lib)
+            sed = which('sed')
+            sed('-i', 's#-dll .*#-dll %s#' % prefix.lib.join('libnrnmech.so'), prefix.bin.special)
```
Did you get library not found errors without this?

ferdonline replied:
Just in case we, at some point, we build in shared lib mode (discussion that maybe someday the lib should always be created and used by special)

pramodk commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -83,41 +94,78 @@ def syn2_dep_libs(self):
-    def check_install(self):
-        special = '%s/special' % join_path(self.prefix, os.path.basename(self.neuron_archdir))
-        if not os.path.isfile(special):
-            raise RuntimeError("Neurodamus installion check failed, %s not created!" % special)
```
I wonder if we should keep above check. This has been useful in multiple occasions for me (e.g. when nrnivmodl finish without error code)

pramodk commented on var/spack/repos/builtin/packages/neurodamus/package.py:
```diff
@@ -83,41 +94,78 @@ def syn2_dep_libs(self):
+                if m.name == 'PYTHONPATH':
+                    run_env.prepend_path('PYTHONPATH', m.value)
+            run_env.prepend_path('PYTHONPATH', self.prefix.python)
+            run_env.set('PYDAMUS', self.prefix.python.join('init.py'))
```
Currently people have : 

```
mpirun ... $HOC_LIBRARY_PATH/init.hoc
```

I like following for consistency:

```
mpirun ... $PYTHON_LIBRARY_PATH/init.py
```

Instead of `mpirun ... $PYDAMUS`. So may be we can set something like PYTHON_LIBRARY_PATH or similar instead of PYDAMUS?

ferdonline replied:
Addressed

pramodk commented:
Please retest.

ferdonline committed:
Several neurodamus improvements

 - syn2 variant renamed to syntool, and made default
 - exporting location of NEURODAMUS_PYTHON, not PYDAMUS
 - more refactoring

ferdonline commented:
please restest

pramodk reviewed (approved):
LGTM
